The ssl_unsupported.h appending in ssl.h causes compilation errors

Currently, the `#include <openssl/ssl_unsupported.h>` is appended to the `ssl.h`, which aims to raise compilation errors when any of APIs specified in `ssl_unsupported.h` is used by an application. However, the way it is appended causes errors even when no APIs are used.
```
// ssl.h
...
#ifndef HEADER_SSL_H
# define HEADER_SSL_H
...

#include <openssl/ssl3.h>
...
#include <openssl/srtp.h>
...

/* Declarations of target APIs to be unsupported. */

#endif
#include <openssl/ssl_unsupported.h>
```

The root cause is that the inclusion of `ssl_unsupported.h` is outside the scope of `#ifndef HEADER_SSL_H` and the `ssl.h` header itself includes `ssl3.h` and `srtp.h`, both of which include `ssl.h`. This causes `ssl_unsupported.h` to be picked up before reaching the declarations of target APIs. See the following trace:

```
. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/opensslconf.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/comp.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/opensslconf.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/crypto.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/comperr.h
.... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/opensslconf.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/crypto.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/hmac.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/opensslconf.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/async.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/libcxx/stdlib.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/asyncerr.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ct.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/opensslconf.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/sslerr.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl2.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl3.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl.h
.... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl_unsupported.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/tls1.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/dtls1.h
.. /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/srtp.h
... /home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl.h
In file included from /home/ssh_office/workspace/oe-ms2/tests/tls_e2e/server_enc/openssl_server.cpp:23:
/home/ssh_office/workspace/oe-ms2/build/output/include/openenclave/openssl/ssl.h:1983:12: error: expected ')'
```

The fix would be simply making the inclusion of `ssl_unsupported.h` inside the scope of `#ifndef HEADER_SSL_H` as follows.
```
// ssl.h
...
#ifndef HEADER_SSL_H
# define HEADER_SSL_H
...

#include <openssl/ssl_unsupported.h>
#endif
```

--- a/3rdparty/openssl/append-unsupported
+++ b/3rdparty/openssl/append-unsupported
@@ -29,14 +29,16 @@ fi
 
 headers=`find ${directory} -name '*.h'`
 
-## Append the inclusion of unsupported.h */
+## Append the inclusion of *_unsupported.h
 for i in ${headers}
 do
     if [[ "${i}" != *"crypto.h" ]] && [[ "${i}" != *"ssl.h" ]] && [[ "${i}" != *"x509_vfy.h" ]]; then
         continue
     fi
     name="$(basename -s .h $i)"
-    echo "#include <openssl/${name}_unsupported.h>" >> ${i}
+    # We assume that every header ends with #endif, which is the case for OpenSSL.
+    sed -i '$d' ${i}
+    echo -e "#include <openssl/${name}_unsupported.h>\n#endif" >> ${i}
     if [ "$?" != "0" ]; then
         echo "$0: failed to append ${string}"
         exit 1

--- a/tests/openssl_unsupported/enc/CMakeLists.txt
+++ b/tests/openssl_unsupported/enc/CMakeLists.txt
@@ -41,21 +41,26 @@ function (add_unsupported_test NAME)
             --config $<CONFIGURATION>
     WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
 
-  # In addition to expecting the compilation to fail, checking whether the log includes
-  # the specific error message, which indicates the unsupported.h header is
-  # correctly included.
-  set_tests_properties(
-    openssl_unsupported_${NAME}
-    PROPERTIES
-      WILL_FAIL
-      TRUE
-      FAIL_REGULAR_EXPRESSION
-      "warning: .* is deprecated: The function|macro may be unsafe inside an enclave"
-  )
+  # The compilation of the baseline case (i.e., not linking against any of unsupported APIs
+  # but simply including the headers) is expected to succeed.
+  if (NOT ${NAME} STREQUAL "baseline")
+    # In addition to expecting the compilation to fail, checking whether the log includes
+    # the specific error message, which indicates the unsupported.h header is
+    # correctly included.
+    set_tests_properties(
+      openssl_unsupported_${NAME}
+      PROPERTIES
+        WILL_FAIL
+        TRUE
+        FAIL_REGULAR_EXPRESSION
+        "warning: .* is deprecated: The function|macro may be unsafe inside an enclave"
+    )
+  endif ()
 
 endfunction (add_unsupported_test)
 
 set(UNSUPPORTED_LIST
+    BASELINE
     OPENSSL_INIT_LOAD_CONFIG
     SSL_CTX_set_default_verify_paths
     SSL_CTX_set_default_verify_dir

